### PR TITLE
TE-10560 Documentation page for Data Import dynamic chunks

### PR DESCRIPTION
**Pull request by yaroslav-spryker**
## PR Description

Documentation page for Data Import dynamic chunks

## Checklist
- [ ] I agree with the Code Contribution License Agreement in CONTRIBUTING.md

### CODE REVIEW

File: docs/scos/dev/guidelines/performance-guidelines/spryker-elastic-computing/data-import-dynamic-chunks.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+description: This guideline explains how to Data Import dynamic chunks feature is implemented.
+last_updated: May 16, 2022
+template: concept-topic-template
+---
+
+# Status quo
+Some PHP-performed long lasting operation (such as data import) may consume considerable amount memory during script execution. In best case scenario, the memory usage keeps a steady level until the end of the execution. In some cases, a significant amount of available memory is not used.
+
+The following chart represents a memory consumption profile for a long lasting PHP operation.
+
```
Comment (yaroslav-spryker) [inline docs/scos/dev/guidelines/performance-guidelines/spryker-elastic-computing/data-import-dynamic-chunks.md:12]: Please add an image

image

File: docs/scos/dev/guidelines/performance-guidelines/spryker-elastic-computing/data-import-dynamic-chunks.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+Using less memory creates the main issue here: operation wall time is increased due to the number of I/O operations. Having more data in memory for processing leads to decreased number of I/O thus to decreased wall time of the operation as well.
+
+# Goal
+
+The goal is to increase a memory consumption and this way to decrease the number of I/O operations and decrease wall time of the operation.
+
+# Solution
+
+In order to use the memory more efficiently new approach was introduced. The batch size now is calculated based on the amount of memory, available from the system which is left for current thread. And only when memory is full by some threshold value database insert is performed.
```
Comment (andriitserkovnyi) [inline docs/scos/dev/guidelines/performance-guidelines/spryker-elastic-computing/data-import-dynamic-chunks.md:32]: Does database insert mean that more memory is allocated to the operation?

File: docs/scos/dev/guidelines/performance-guidelines/spryker-elastic-computing/data-import-dynamic-chunks.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+- Graduality factor
+- Memory threshold percent
+
+## Graduality factor
+
+Graduality factor is a number that defines how many steps a program needs to take before a maximum allowed memory usage is reached.
+
+The larger the gradually factor is, the more approximation steps the program takes, and the higher  the accuracy is. A smaller gradually factor makes application work faster but rougher.
+
+## Memory threshold percent
```
Comment (andriitserkovnyi) [inline docs/scos/dev/guidelines/performance-guidelines/spryker-elastic-computing/data-import-dynamic-chunks.md:47]: percentage?

File: docs/scos/dev/guidelines/performance-guidelines/spryker-elastic-computing/data-import-dynamic-chunks.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+To use the memory more efficiently the batch size is now calculated based on the amount of memory  available for current thread. When memory is full by some threshold value database insert is performed.
+
+The algorithm uses the following variables:
+- Allowed total memory for PHP thread
+- Currently used memory
+- Maximum used memory during the execution of current thread
+
+Implementation is based on a gradual principle when we reach a memory threshold usage step-by-step.
+
+Two variables are also important here:
```
Comment (andriitserkovnyi) [inline docs/scos/dev/guidelines/performance-guidelines/spryker-elastic-computing/data-import-dynamic-chunks.md:37]: Why are these variables not listed together with the variables above?

File: docs/scos/dev/guidelines/performance-guidelines/spryker-elastic-computing/data-import-dynamic-chunks.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+
+
+For bulk data imports we used to define batch size, like 2000 or 5000 items. On the prior image, memory consumption reaches the level that corresponds to the batch size, which leaves a lot of  unused memory.
+
+Using less memory generates more I/O operations, which subsequently increases the operation's processing time. Having more data in memory for processing decreased number of I/O and the processing time.
+
+## Dynamic memory allocation
+
+To use the memory more efficiently the batch size is now calculated based on the amount of memory  available for current thread. When memory is full by some threshold value database insert is performed.
```
Comment (andriitserkovnyi) [inline docs/scos/dev/guidelines/performance-guidelines/spryker-elastic-computing/data-import-dynamic-chunks.md:28]: Does database insert mean that more memory is allocated to the operation?

**Review by andriitserkovnyi — APPROVED** [review]

**Pull request merged**